README
Digest::SHA3 version 0.23
=========================

Digest::SHA3 is a complete implementation of the NIST SHA-3 cryptographic
hash function, known originally as Keccak.  It gives Perl programmers a
convenient way to calculate SHA3-224, SHA3-256, SHA3-384, and SHA3-512
message digests, as well as variable-length hashes using SHAKE128
and SHAKE256.  The module can handle all types of input, including
partial-byte data.

Digest::SHA3 is written in C for speed.  If your platform lacks a C
compiler, perhaps you can find the module in a binary form compatible
with your particular processor and operating system.

The initial version of Digest::SHA3 was written to emphasize simplicity
and correctness.  It passed all known-answer-tests for short and long
messages.  No particular emphasis was placed on performance.  However,
as the SHA-3 standard matures and achieves 'final' status, versions of
this module will endeavor to perform on a par with optimized portable
implementations.

The module omits all HMAC-SHA-3 functions until such time as official
test vectors become available.

The tests subdirectory (t/*.t) contains a set of test vectors compiled
from both official and informal sources.

INSTALLATION

To install this module type the following:

	perl Makefile.PL
	make
	make test
	make install

DEPENDENCIES

	None

COPYRIGHT AND LICENSE

Copyright (C) 2012-2015 Mark Shelor

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

Please refer to the Perl Artistic License for details:

http://search.cpan.org/perldoc?perlartistic


Changes
Revision history for Perl extension Digest::SHA3.

0.23  Sun Jan  4 05:36:30 MST 2015
	- updated to reflect Draft FIPS 202
		-- append domain separation bits to message
		-- implement SHAKE128 and SHAKE256 Extendable-Output
			Functions (XOFs)

0.22  Sun Jun  1 00:15:46 MST 2014
	- fixed reserved-word clash when compiling with C++
		-- use 'classname' instead of 'class'
			-- ref. SHA3.xs (rt.cpan.org #96090)

0.21  Fri May 16 10:21:46 MST 2014
	- restored original 'addfile' for use on opened file handles
		-- allows callbacks in place of actual files
			-- ref. IO::Callback (rt.cpan.org #95643)
	- re-established inheritance from Digest::base
		-- to pick up future Digest enhancements automatically
	- cleaned up documentation

0.20  Wed May  7 07:57:10 MST 2014
	- consolidated all dynamic memory allocation into XSUBs
		-- streamlines referencing of SHA3 objects
		-- simplifies DESTROYing of objects
	- enhanced Makefile.PL to allow 'use warnings'
		-- automatically reverts to $^W for early Perls
	- scrubbed C and Perl code to remove all compiler warnings

0.12  Sat Apr 19 05:14:50 MST 2014
	- added universal newlines mode ("U") to addfile and sha3sum
		-- based on Python Universal Newlines concept
			-- newlines identical across MacOS, DOS, and UNIX
		-- will deprecate portable mode ("p") in future
			-- "U" mode is cleaner and more efficient
	- enhanced performance
		-- reduced number of dynamic memory allocations
		-- sped up addfile method with use of C code
			-- ref. SHA3.xs (_addfilebin and _addfileuniv)
	- changed text file test (-T) to act on filehandles
		-- improves consistency when reading from STDIN
		-- still must act on filenames for early Perls (< 5.6)
	- sealed memory leak in SHA3.xs
		-- arose only with SvPVbyte exceptions during eval
	- patched inheritence bug (ref: rt.cpan.org #94830)
		-- use sv_isobject/sv_derived_from instead of sv_isa
	- added 'allocated' flag to SHA3 structure (ref. src/sha3.h)
		-- to guard against Perl double frees

0.11  Mon Feb 17 16:42:04 MST 2014
	- tightened code in SHA3.xs
		-- added sv_isa checks when invoking methods

0.10  Thu Jan 30 08:24:30 MST 2014
	- improved the performance of hexadecimal output functions
		-- ref. 'shahex' in src/sha3.c
		-- thanks to Thomas Drugeon for ideas and test script

0.09  Sun Jan  5 19:08:32 MST 2014
	- added a 'squeeze' method for SHA3-0 objects
		-- to construct SHA3-0 digests of any desired length

0.08  Wed Jun 26 04:32:06 MST 2013
	- workaround for repeated calls to shaclose
		-- ref. Bug #86295 (posted at Digest::SHA)
		-- need to explicitly reset internal pointer to NULL
			ref. shaclose() in SHA3.xs
	- corrected typos in sha3sum script
		-- ref. Bug #85430 (posted at Digest::SHA)

0.07  Sat Mar  9 17:36:14 MST 2013
	- untweaked Makefile.PL to remove dependencies of SHA3.c
		-- dependencies were breaking builds on VMS
		-- retaining dependencies provides too little benefit
			for cost of portable workaround

0.06  Mon Mar  4 08:12:04 MST 2013
	- removed code for standalone C operation (no longer used)
		-- eliminates need for external symbols
		-- reduces size of object files
		-- thanks to Marc Lehmann for suggestions
	- tweaked Makefile.PL to show dependencies of SHA3.c

0.05  Thu Jan 24 04:54:14 MST 2013
	- accommodated Unicode string input
		-- by using SvPVbyte instead of SvPV in SHA3.xs
			-- provided workaround for Perl 5.6
		-- added new test script t/unicode.t
	- provided documentation to describe Unicode handling
	- obtained slight speedup on Intel/gcc
		-- by setting -O1 and -fomit-frame-pointer
	- adopted (from Digest::SHA) workaround for DEC compiler bug

0.04  Sun Nov 11 19:20:06 MST 2012
	- enhanced performance
		-- typically 10-15% faster than 0.03
		-- code is still easy to read and modify
	- filled in a few documentation omissions

0.03  Mon Oct 29 04:01:06 MST 2012
	- attained a 2.5x performance increase through loop unrolling
		-- code is still easy to follow
			-- clearly traceable to Keccak pseudo code

0.02  Thu Oct 25 19:18:58 MST 2012
	- original version: adapted from Digest::SHA version 5.72


src/sha3.c
/*
 * sha3.c: routines to compute SHA-3 digests
 *
 * Ref: http://keccak.noekeon.org/specs_summary.html
 *
 * Copyright (C) 2012-2015 Mark Shelor, All Rights Reserved
 *
 * Version: 0.23
 * Sun Jan  4 05:36:30 MST 2015
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include "sha3.h"

#define UCHR	unsigned char		/* useful abbreviations */
#define UINT	unsigned int
#define ULNG	unsigned long
#define W64	SHA64
#define C64	SHA64_CONST
#define SR64	SHA64_SHR
#define SL64	SHA64_SHL

/* word2mem: write 64-bit value in little-endian order */
static void word2mem(UCHR *mem, W64 w)
{
	int i;
	UCHR *p = mem;

	for (i = 0; i < 8; i++, w >>= 8)
		*p++ = (UCHR) (w & 0xff);
}

static W64 RC[] = {	/* Keccak round constants */
	C64(0x0000000000000001), C64(0x0000000000008082),
	C64(0x800000000000808a), C64(0x8000000080008000),
	C64(0x000000000000808b), C64(0x0000000080000001),
	C64(0x8000000080008081), C64(0x8000000000008009),
	C64(0x000000000000008a), C64(0x0000000000000088),
	C64(0x0000000080008009), C64(0x000000008000000a),
	C64(0x000000008000808b), C64(0x800000000000008b),
	C64(0x8000000000008089), C64(0x8000000000008003),
	C64(0x8000000000008002), C64(0x8000000000000080),
	C64(0x000000000000800a), C64(0x800000008000000a),
	C64(0x8000000080008081), C64(0x8000000000008080),
	C64(0x0000000080000001), C64(0x8000000080008008)
};

/* ROTL: rotate 64-bit word left by n bit positions */
#define ROTL(w, n) (SR64((w), (64 - (n))) | SL64((w), (n)))

/* keccak_f: apply KECCAK-f[1600] permutation for 24 rounds */
static void keccak_f(W64 A[][5])
{
	int i;
	W64 *rc = RC;
	for (i = 0; i < 24; i++, rc++) {
		W64 B[5][5], C[5], D[5];
		C[0] = A[0][0]^A[0][1]^A[0][2]^A[0][3]^A[0][4];
		C[1] = A[1][0]^A[1][1]^A[1][2]^A[1][3]^A[1][4];
		C[2] = A[2][0]^A[2][1]^A[2][2]^A[2][3]^A[2][4];
		C[3] = A[3][0]^A[3][1]^A[3][2]^A[3][3]^A[3][4];
		C[4] = A[4][0]^A[4][1]^A[4][2]^A[4][3]^A[4][4];
		D[0] = C[4] ^ ROTL(C[1], 1);
		D[1] = C[0] ^ ROTL(C[2], 1);
		D[2] = C[1] ^ ROTL(C[3], 1);
		D[3] = C[2] ^ ROTL(C[4], 1);
		D[4] = C[3] ^ ROTL(C[0], 1);
		A[0][0] ^= D[0];
		A[0][1] ^= D[0];
		A[0][2] ^= D[0];
		A[0][3] ^= D[0];
		A[0][4] ^= D[0];
		A[1][0] ^= D[1];
		A[1][1] ^= D[1];
		A[1][2] ^= D[1];
		A[1][3] ^= D[1];
		A[1][4] ^= D[1];
		A[2][0] ^= D[2];
		A[2][1] ^= D[2];
		A[2][2] ^= D[2];
		A[2][3] ^= D[2];
		A[2][4] ^= D[2];
		A[3][0] ^= D[3];
		A[3][1] ^= D[3];
		A[3][2] ^= D[3];
		A[3][3] ^= D[3];
		A[3][4] ^= D[3];
		A[4][0] ^= D[4];
		A[4][1] ^= D[4];
		A[4][2] ^= D[4];
		A[4][3] ^= D[4];
		A[4][4] ^= D[4];
		B[0][0] = A[0][0];
		B[1][3] = ROTL(A[0][1], 36);
		B[2][1] = ROTL(A[0][2], 3);
		B[3][4] = ROTL(A[0][3], 41);
		B[4][2] = ROTL(A[0][4], 18);
		B[0][2] = ROTL(A[1][0], 1);
		B[1][0] = ROTL(A[1][1], 44);
		B[2][3] = ROTL(A[1][2], 10);
		B[3][1] = ROTL(A[1][3], 45);
		B[4][4] = ROTL(A[1][4], 2);
		B[0][4] = ROTL(A[2][0], 62);
		B[1][2] = ROTL(A[2][1], 6);
		B[2][0] = ROTL(A[2][2], 43);
		B[3][3] = ROTL(A[2][3], 15);
		B[4][1] = ROTL(A[2][4], 61);
		B[0][1] = ROTL(A[3][0], 28);
		B[1][4] = ROTL(A[3][1], 55);
		B[2][2] = ROTL(A[3][2], 25);
		B[3][0] = ROTL(A[3][3], 21);
		B[4][3] = ROTL(A[3][4], 56);
		B[0][3] = ROTL(A[4][0], 27);
		B[1][1] = ROTL(A[4][1], 20);
		B[2][4] = ROTL(A[4][2], 39);
		B[3][2] = ROTL(A[4][3], 8);
		B[4][0] = ROTL(A[4][4], 14);
		A[0][0] = B[0][0] ^ (~B[1][0] & B[2][0]);
		A[0][1] = B[0][1] ^ (~B[1][1] & B[2][1]);
		A[0][2] = B[0][2] ^ (~B[1][2] & B[2][2]);
		A[0][3] = B[0][3] ^ (~B[1][3] & B[2][3]);
		A[0][4] = B[0][4] ^ (~B[1][4] & B[2][4]);
		A[1][0] = B[1][0] ^ (~B[2][0] & B[3][0]);
		A[1][1] = B[1][1] ^ (~B[2][1] & B[3][1]);
		A[1][2] = B[1][2] ^ (~B[2][2] & B[3][2]);
		A[1][3] = B[1][3] ^ (~B[2][3] & B[3][3]);
		A[1][4] = B[1][4] ^ (~B[2][4] & B[3][4]);
		A[2][0] = B[2][0] ^ (~B[3][0] & B[4][0]);
		A[2][1] = B[2][1] ^ (~B[3][1] & B[4][1]);
		A[2][2] = B[2][2] ^ (~B[3][2] & B[4][2]);
		A[2][3] = B[2][3] ^ (~B[3][3] & B[4][3]);
		A[2][4] = B[2][4] ^ (~B[3][4] & B[4][4]);
		A[3][0] = B[3][0] ^ (~B[4][0] & B[0][0]);
		A[3][1] = B[3][1] ^ (~B[4][1] & B[0][1]);
		A[3][2] = B[3][2] ^ (~B[4][2] & B[0][2]);
		A[3][3] = B[3][3] ^ (~B[4][3] & B[0][3]);
		A[3][4] = B[3][4] ^ (~B[4][4] & B[0][4]);
		A[4][0] = B[4][0] ^ (~B[0][0] & B[1][0]);
		A[4][1] = B[4][1] ^ (~B[0][1] & B[1][1]);
		A[4][2] = B[4][2] ^ (~B[0][2] & B[1][2]);
		A[4][3] = B[4][3] ^ (~B[0][3] & B[1][3]);
		A[4][4] = B[4][4] ^ (~B[0][4] & B[1][4]);
		A[0][0] ^= *rc;
	}
}

/* sha3: update SHA3 state with one block of data */
static void sha3(SHA3 *s, UCHR *block)
{
	unsigned int i, x, y;
	W64 P0[5][5];

	for (i = 0; i < s->blocksize/64; i++, block += 8)
		MEM2WORD(&P0[i%5][i/5], block);
	for (x = 0; x < 5; x++)
		for (y = 0; y < 5; y++) {
			if (x + y*5 >= s->blocksize/64)
				break;
			s->S[x][y] ^= P0[x][y];
		}
	keccak_f(s->S);
}

/* digcpy: write SHA3 state to digest buffer */
static UCHR *digcpy(SHA3 *s)
{
	unsigned int x, y;
	UCHR *Z = s->digest;
	int outbits = s->digestlen*8;

	while (outbits > 0) {
		for (y = 0; y < 5; y++)
			for (x = 0; x < 5; x++, Z += 8) {
				if (x + y*5 >= s->blocksize/64)
					break;
				word2mem(Z, s->S[x][y]);
			}
		if ((outbits -= s->blocksize) > 0)
			keccak_f(s->S);
	}
	return(s->digest);
}

#define NBYTES(nbits)   (((nbits) + 7) >> 3)
#define HEXLEN(nbytes)	((nbytes) << 1)
#define B64LEN(nbytes)	(((nbytes) % 3 == 0) ? ((nbytes) / 3) * 4 \
			: ((nbytes) / 3) * 4 + ((nbytes) % 3) + 1)

#define SHA3_INIT(s, algo, xof)					\
	do {							\
		Zero(s, 1, SHA3);				\
		s->alg = algo;					\
		s->shake = xof;					\
		s->blocksize = algo ## _BLOCK_BITS;		\
		s->digestlen = algo ## _DIGEST_BITS >> 3;	\
	} while (0)

/* sharewind: resets digest object */
static void sharewind(SHA3 *s)
{
	if      (s->alg == SHA3_224) SHA3_INIT(s, SHA3_224, 0);
	else if (s->alg == SHA3_256) SHA3_INIT(s, SHA3_256, 0);
	else if (s->alg == SHA3_384) SHA3_INIT(s, SHA3_384, 0);
	else if (s->alg == SHA3_512) SHA3_INIT(s, SHA3_512, 0);
	else if (s->alg == SHAKE128) SHA3_INIT(s, SHAKE128, 1);
	else if (s->alg == SHAKE256) SHA3_INIT(s, SHAKE256, 1);
}

/* shainit: initializes digest object */
static int shainit(SHA3 *s, int alg)
{
	if (alg != SHA3_224 && alg != SHA3_256 &&
		alg != SHA3_384 && alg != SHA3_512 &&
		alg != SHAKE128 && alg != SHAKE256)
		return 0;
	s->alg = alg;
	sharewind(s);
	return 1;
}

/* shadirect: updates state directly (w/o going through s->block) */
static ULNG shadirect(UCHR *bitstr, ULNG bitcnt, SHA3 *s)
{
	ULNG savecnt = bitcnt;

	while (bitcnt >= s->blocksize) {
		sha3(s, bitstr);
		bitstr += (s->blocksize >> 3);
		bitcnt -= s->blocksize;
	}
	if (bitcnt > 0) {
		Copy(bitstr, s->block, NBYTES(bitcnt), char);
		s->blockcnt = bitcnt;
	}
	return(savecnt);
}

/* shabytes: updates state for byte-aligned data in s->block */
static ULNG shabytes(UCHR *bitstr, ULNG bitcnt, SHA3 *s)
{
	UINT offset;
	UINT nbits;
	ULNG savecnt = bitcnt;

	offset = s->blockcnt >> 3;
	if (s->blockcnt + bitcnt >= s->blocksize) {
		nbits = s->blocksize - s->blockcnt;
		Copy(bitstr, s->block+offset, nbits>>3, char);
		bitcnt -= nbits;
		bitstr += (nbits >> 3);
		sha3(s, s->block), s->blockcnt = 0;
		shadirect(bitstr, bitcnt, s);
	}
	else {
		Copy(bitstr, s->block+offset, NBYTES(bitcnt), char);
		s->blockcnt += bitcnt;
	}
	return(savecnt);
}

/* shabits: updates state for bit-aligned data in s->block */
static ULNG shabits(UCHR *bitstr, ULNG bitcnt, SHA3 *s)
{
	UINT i;
	UINT gap;
	ULNG nbits;
	UCHR buf[1<<9];
	UINT bufsize = sizeof(buf);
	ULNG bufbits = (ULNG) bufsize << 3;
	UINT nbytes = NBYTES(bitcnt);
	ULNG savecnt = bitcnt;

	gap = 8 - s->blockcnt % 8;
	s->block[s->blockcnt>>3] &= ~0 << gap;
	s->block[s->blockcnt>>3] |= *bitstr >> (8 - gap);
	s->blockcnt += bitcnt < gap ? bitcnt : gap;
	if (bitcnt < gap)
		return(savecnt);
	if (s->blockcnt == s->blocksize)
		sha3(s, s->block), s->blockcnt = 0;
	if ((bitcnt -= gap) == 0)
		return(savecnt);
	while (nbytes > bufsize) {
		for (i = 0; i < bufsize; i++)
			buf[i] = bitstr[i] << gap | bitstr[i+1] >> (8-gap);
		nbits = bitcnt < bufbits ? bitcnt : bufbits;
		shabytes(buf, nbits, s);
		bitcnt -= nbits, bitstr += bufsize, nbytes -= bufsize;
	}
	for (i = 0; i < nbytes - 1; i++)
		buf[i] = bitstr[i] << gap | bitstr[i+1] >> (8-gap);
	buf[nbytes-1] = bitstr[nbytes-1] << gap;
	shabytes(buf, bitcnt, s);
	return(savecnt);
}

/* shawrite: triggers a state update using data in bitstr/bitcnt */
static ULNG shawrite(UCHR *bitstr, ULNG bitcnt, SHA3 *s)
{
	if (bitcnt < 1)
		return(0);
	if (s->blockcnt == 0)
		return(shadirect(bitstr, bitcnt, s));
	else if (s->blockcnt % 8 == 0)
		return(shabytes(bitstr, bitcnt, s));
	else
		return(shabits(bitstr, bitcnt, s));
}

#define SETBIT(s, pos)  s[(pos) >> 3] |= (UCHR)  (0x01 << ((pos) % 8))
#define CLRBIT(s, pos)  s[(pos) >> 3] &= (UCHR) ~(0x01 << ((pos) % 8))

/* shafinish: pads remaining block(s) and computes final digest state */
static void shafinish(SHA3 *s)
{
	UCHR domain = s->shake ? 0x1f : 0x06;

	if (s->padded)
		return;
	s->padded = 1;
	if (s->blockcnt % 8 == 0) {
		s->block[s->blockcnt/8] = domain;
		s->blockcnt += 8;
		while (s->blockcnt < s->blocksize)
			s->block[s->blockcnt/8] = 0x00, s->blockcnt += 8;
		s->block[(s->blockcnt/8)-1] |= 0x80;
		sha3(s, s->block);
		return;
	}
	s->block[NBYTES(s->blockcnt)-1] >>= (8 - (s->blockcnt % 8));
	for (; domain; domain >>= 1) {
		if (domain & 0x01)
			SETBIT(s->block, s->blockcnt), s->blockcnt++;
		else
			CLRBIT(s->block, s->blockcnt), s->blockcnt++;
		if (s->blockcnt == s->blocksize)
			sha3(s, s->block), s->blockcnt = 0;
	}
	while (s->blockcnt < (s->blocksize - 1))
		CLRBIT(s->block, s->blockcnt), s->blockcnt++;
	SETBIT(s->block, s->blockcnt), s->blockcnt++;
	sha3(s, s->block);
}

/* shasqueeze: returns pointer to squeezed digest (binary) */
static UCHR *shasqueeze(SHA3 *s)
{
	if (s->alg != SHAKE128 && s->alg != SHAKE256)
		return(NULL);
	digcpy(s);
	keccak_f(s->S);
	return(s->digest);
}

#define shadigest(state)	digcpy(state)

/* xmap: translation map for hexadecimal encoding */
static char xmap[] =
	"0123456789abcdef";

/* shahex: returns pointer to current digest (hexadecimal) */
static char *shahex(SHA3 *s)
{
	int i;
	char *h;
	UCHR *d;

	d = digcpy(s);
	s->hex[0] = '\0';
	if (HEXLEN((size_t) s->digestlen) >= sizeof(s->hex))
		return(s->hex);
	for (i = 0, h = s->hex; i < s->digestlen; i++) {
		*h++ = xmap[(*d >> 4) & 0x0f];
		*h++ = xmap[(*d++   ) & 0x0f];
	}
	*h = '\0';
	return(s->hex);
}

/* bmap: translation map for Base 64 encoding */
static char bmap[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

/* encbase64: encodes input (0 to 3 bytes) into Base 64 */
static void encbase64(UCHR *in, int n, char *out)
{
	UCHR byte[3] = {0, 0, 0};

	out[0] = '\0';
	if (n < 1 || n > 3)
		return;
	Copy(in, byte, n, UCHR);
	out[0] = bmap[byte[0] >> 2];
	out[1] = bmap[((byte[0] & 0x03) << 4) | (byte[1] >> 4)];
	out[2] = bmap[((byte[1] & 0x0f) << 2) | (byte[2] >> 6)];
	out[3] = bmap[byte[2] & 0x3f];
	out[n+1] = '\0';
}

/* shabase64: returns pointer to current digest (Base 64) */
static char *shabase64(SHA3 *s)
{
	int n;
	UCHR *q;
	char out[5];

	q = digcpy(s);
	s->base64[0] = '\0';
	if (B64LEN((size_t) s->digestlen) >= sizeof(s->base64))
		return(s->base64);
	for (n = s->digestlen; n > 3; n -= 3, q += 3) {
		encbase64(q, 3, out);
		strcat(s->base64, out);
	}
	encbase64(q, n, out);
	strcat(s->base64, out);
	return(s->base64);
}


Makefile.PL
require 5.003000;

use strict;
use ExtUtils::MakeMaker;
use Getopt::Std;
use Config qw(%Config);

use vars qw($opt_w);

my $PM = 'lib/Digest/SHA3.pm';
my $SHA3SUM = 'sha3sum';

	# 'use warnings' if possible, but stay portable

my($use_warnings, @EDITs);
my $warn_1 = 'BEGIN { $^W = 1 }';
my $warn_0 = 'BEGIN { $^W = 0 }';
{
	eval "require warnings; import warnings";
	$use_warnings = $@ ? 0 : 1;

	local(@ARGV) = ($PM, $SHA3SUM);
	while (<>) {
		if (
		(!$use_warnings && /^(use|no) warnings\b/) ||
		( $use_warnings && /^\Q$warn_1\E # use warnings\b/) ||
		( $use_warnings && /^\Q$warn_0\E # no warnings\b/)) {
			push @EDITs, $ARGV;
			close ARGV;
		}
	}
}

if (@EDITs) {
	local($^I, @ARGV) = ('', @EDITs);
	while (<>) {
		if ($use_warnings) {
			s/^\Q$warn_1\E # (.*)$/$1/;
			s/^\Q$warn_0\E # (.*)$/$1/;
		}
		else {
			s/^(use warnings\b.*)$/$warn_1 # $1/;
			s/^(no warnings\b.*)$/$warn_0 # $1/;
		}
		print;
	}
}


getopts('w');

my @extra = ();

	# Workaround for DEC compiler bug, adapted from Digest::MD5

if ($^O eq 'VMS') {
	if (defined($Config{ccname})) {
		if (grep(/VMS_VAX/, @INC) && ($Config{ccname} eq 'DEC')) {
			# VAX optimizer, even up to v6.4, gets stuck
			push(@extra, OPTIMIZE => "/Optimize=(NODISJOINT)");
		}
	}
}

	# Enhance performance on Intel when using gcc

if ($Config{archname} =~ /^i[3456]86/ && $Config{ccname} eq 'gcc') {
	push(@extra, OPTIMIZE => '-O1 -fomit-frame-pointer');
}

push(@extra, CCFLAGS => '-W -Wall ' . $Config{ccflags}) if $opt_w;

my %attr = (
	'NAME'		=> 'Digest::SHA3',
	'VERSION_FROM'	=> $PM,
	'LIBS'		=> [''],
	'INC'		=> '-I.',
	'EXE_FILES'	=> [ $SHA3SUM ],
	@extra,
);

my $MMversion = $ExtUtils::MakeMaker::VERSION || '0.00_00';
$attr{NO_META} = 1 if $MMversion ge '6.10_03';

WriteMakefile(%attr);


src/sha3.h
/*
 * sha3.h: header file for SHA-3 routines
 *
 * Ref: http://keccak.noekeon.org/specs_summary.html
 *
 * Copyright (C) 2012-2015 Mark Shelor, All Rights Reserved
 *
 * Version: 0.23
 * Sun Jan  4 05:36:30 MST 2015
 *
 */

#ifndef _INCLUDE_SHA3_H_
#define _INCLUDE_SHA3_H_

#include <limits.h>

#define SHA64_SHR(x, n)	((x) >> (n))
#define SHA64_SHL(x, n)	((x) << (n))

#define SHA64_ALIGNED

#if defined(ULONG_LONG_MAX) || defined(ULLONG_MAX) || defined(HAS_LONG_LONG)
	#define SHA_ULL_EXISTS
#endif

#if (((ULONG_MAX >> 16) >> 16) >> 16) >> 15 == 1UL
	#define SHA64	unsigned long
	#define SHA64_CONST(c)	c ## UL
#elif defined(SHA_ULL_EXISTS) && defined(LONGLONGSIZE) && LONGLONGSIZE == 8
	#define SHA64	unsigned long long
	#define SHA64_CONST(c)	c ## ULL
#elif defined(SHA_ULL_EXISTS)
	#undef  SHA64_ALIGNED
	#undef  SHA64_SHR
	#define SHA64_MAX	18446744073709551615ULL
	#define SHA64_SHR(x, n)	(((x) & SHA64_MAX) >> (n))
	#define SHA64	unsigned long long
	#define SHA64_CONST(c)	c ## ULL

	/* The following cases detect compilers that
	 * support 64-bit types in a non-standard way */

#elif defined(_MSC_VER)					/* Microsoft C */
	#define SHA64	unsigned __int64
	#define SHA64_CONST(c)	(SHA64) c
#endif

#if defined(BYTEORDER) && ((BYTEORDER==0x1234) || (BYTEORDER==0x12345678))
	#if defined(SHA64_ALIGNED)
		#define MEM2WORD(W, m)	Copy(m, W, 8, char)
	#endif
#endif

#if !defined(MEM2WORD)
	#define MEM2WORD(W, m) *(W) = 				\
		(SHA64) m[0] << 0  | (SHA64) m[1] << 8  |	\
		(SHA64) m[2] << 16 | (SHA64) m[3] << 24 |	\
		(SHA64) m[4] << 32 | (SHA64) m[5] << 40 |	\
		(SHA64) m[6] << 48 | (SHA64) m[7] << 56
#endif

#define SHA3_224	224
#define SHA3_256	256
#define SHA3_384	384
#define SHA3_512	512
#define SHAKE128	128000
#define SHAKE256	256000

#define SHA3_224_BLOCK_BITS	1152
#define SHA3_256_BLOCK_BITS	1088
#define SHA3_384_BLOCK_BITS	832
#define SHA3_512_BLOCK_BITS	576
#define SHAKE128_BLOCK_BITS	1344
#define SHAKE256_BLOCK_BITS	1088

#define SHA3_224_DIGEST_BITS	224
#define SHA3_256_DIGEST_BITS	256
#define SHA3_384_DIGEST_BITS	384
#define SHA3_512_DIGEST_BITS	512
#define SHAKE128_DIGEST_BITS	SHAKE128_BLOCK_BITS
#define SHAKE256_DIGEST_BITS	SHAKE256_BLOCK_BITS

#define SHA3_MAX_BLOCK_BITS	SHAKE128_BLOCK_BITS
#define SHA3_MAX_DIGEST_BITS	SHAKE128_DIGEST_BITS
#define SHA3_MAX_HEX_LEN	(SHA3_MAX_DIGEST_BITS / 4)
#define SHA3_MAX_BASE64_LEN	(1 + (SHA3_MAX_DIGEST_BITS / 6))

typedef struct SHA3 {
	int alg;
	SHA64 S[5][5];
	unsigned char block[SHA3_MAX_BLOCK_BITS/8];
	unsigned int blockcnt;
	unsigned int blocksize;
	unsigned char digest[SHA3_MAX_DIGEST_BITS/8];
	int digestlen;
	char hex[SHA3_MAX_HEX_LEN+1];
	char base64[SHA3_MAX_BASE64_LEN+1];
	int padded;
	int shake;
} SHA3;

#endif	/* _INCLUDE_SHA3_H_ */
